src/Microsoft.OData.Core/ODataParameterReaderCore.cs

```diff
@@ -444,36 +466,44 @@ private static string GetCreateReaderMethodName(ODataParameterReaderState state)
 
         /// <summary>
-        /// Verifies that one of CreateResourceReader(), CreateResourceSetReader() or CreateCollectionReader() can be called.
+        /// Catch any exception thrown by the action passed in; in the exception case move the reader into
+        /// state ExceptionThrown and then rethrow the exception.
         /// </summary>
-        /// <param name="expectedState">The expected state of the reader.</param>
-        private void VerifyCanCreateSubReader(ODataParameterReaderState expectedState)
+        /// <typeparam name="T">The type returned from the <paramref name="action"/> to execute.</typeparam>
+        /// <param name="action">The action to execute.</param>
+        /// <returns>The result of executing the <paramref name="action"/>.</returns>
+        private T InterceptException<T>(Func<ODataParameterReaderCore, T> action)
         {
-            this.inputContext.VerifyNotDisposed();
-            if (this.State != expectedState)
+            try
             {
-                throw new ODataException(Strings.ODataParameterReaderCore_InvalidCreateReaderMethodCalledForState(ODataParameterReaderCore.GetCreateReaderMethodName(expectedState), this.State));
+                return action(this);
             }
-
-            if (this.subReaderState != SubReaderState.None)
+            catch (Exception e)
             {
-                Debug.Assert(this.Name != null, "this.Name != null");
-                throw new ODataException(Strings.ODataParameterReaderCore_CreateReaderAlreadyCalled(ODataParameterReaderCore.GetCreateReaderMethodName(expectedState), this.Name));
+                if (ExceptionUtils.IsCatchableExceptionType(e))
+                {
+                    this.EnterScope(ODataParameterReaderState.Exception, null, null);
+                }
+
+                throw;
             }
         }
 
         /// <summary>
         /// Catch any exception thrown by the action passed in; in the exception case move the reader into
         /// state ExceptionThrown and then rethrow the exception.
         /// </summary>
-        /// <typeparam name="T">The type returned from the <paramref name="action"/> to execute.</typeparam>
+        /// <typeparam name="TResult">The type returned from the <paramref name="action"/></typeparam>
         /// <param name="action">The action to execute.</param>
-        /// <returns>The result of executing the <paramref name="action"/>.</returns>
-        private T InterceptException<T>(Func<T> action)
+        /// <returns>
+        /// A task that represents the asynchronous operation.
+        /// The value of the TResult parameter contains the result of executing the <paramref name="action"/>.
+        /// </returns>
+        private async Task<TResult> InterceptExceptionAsync<TResult>(Func<ODataParameterReaderCore, Task<TResult>> action)
         {
             try
             {
-                return action();
+                return await action(this).ConfigureAwait(false);

```

Comment:
Can we elide the await here?

Reply:
@KenitoInc We can't elide here since we want the exception that is rethrown in the `catch` block to be stashed in the returned task. That won't happen if we elide async/await